lite-roleplay-bot.py

```diff
@@ -0,0 +1,197 @@
+        content = message.content
+        if not isinstance(channel, discord.channel.DMChannel):
+            for mention in re.finditer(r'<@!?(\d+)>', content):
+                user_id = int(mention.group(1))
+                member = discord.utils.get(message.guild.members, id=user_id)
+                if member:
+                    content = content.replace(mention.group(0), f"@{member.name}")
+        messages.append((message.author.display_name, content))
+    return messages
+
+async def background_task():
+    executor = ThreadPoolExecutor(max_workers=1)
+    loop = asyncio.get_running_loop()
+    print("Task Started. Waiting for inputs.")
+    while True:
+        msg_pair: tuple[discord.Message, discord.Message, list] = await queue.get()
+        msg, past, past_messages = msg_pair
+
+        message_content = msg.author.display_name + ": " + replace_mentions_with_usernames(msg.content, msg)
+        past_content = None
+        if past:
+            past_user = past.author.display_name
+            past_content = past_user + ": " + replace_mentions_with_usernames(past.content, past)
+        text = generate_prompt(message_content, past_content, past_messages)
+        response = await loop.run_in_executor(executor, sync_task, text)
+        print(f"Response: {text}\n{response}")
+        
+        if not response.strip():
+            response = "Sorry, I didn't understand that.  Could you please clarify your statement."
+
+        try:
+            await msg.reply(response, mention_author=False)
+        except discord.errors.Forbidden:
+            print("Error: Missing Permissions")
+            await msg.channel.send("Retry")
+
+def sync_task(message):
+    global chatbot
+    input_ids = chatbot.model.tokenize(message)
+    print(f"Input tokens: {input_ids}")  # Log input tokens
+
+    generated_ids = chatbot.model.generate(input_ids, repetition_penalty=1.1, temperature=0.28, top_p=0.95, top_k=40, reset=False)
+
+    # To get the tokens generated by the bot
+    generated_tokens = []
+
+    for token in generated_ids:
+        generated_tokens.append(token)
+    
+    print(f"Generated tokens: {generated_tokens}")  # Log generated tokens
+
+    response = chatbot.model.detokenize(generated_tokens).replace("</s>", "")
+    return response
+
+def generate_prompt(text, pastMessage, past_messages, character_json_path=character):
+    global chatbot
+    max_token_limit = 2000
+    chat_history = ""
+    token_count = 0
+
+    with open(character_json_path, 'r') as f:
+        character_data = json.load(f)
+
+    name = character_data.get('name', '')
+    background = character_data.get('description', '')
+    personality = character_data.get('personality', '')
+    circumstances = character_data.get('world_scenario', '')
+    common_greeting = character_data.get('first_mes', '')
+    past_dialogue = character_data.get('mes_example', '')
+    past_dialogue_formatted = past_dialogue
+
+    for username, message in past_messages:
```

Comment:
are the messages in past_messages sorted from newest to oldest? if not,     newer messages will be cut out when truncating the messages on line 164

Reply:
To be honest, that bit is code I inherited unchanged from @teknium1's original implementation.

It is not relevant either to the `ctransformers` port or to making the bot work with quantized ggml based models.

I might fix it at some point if my use case requires it.

In fact, I think the latest version of my code has actually dropped the `past_dialogue_formatted` stuff from the `lite-roleplay-bot.py`.

Reply:
Yep, I've removed all the `past_dialogue` stuff entirely.

Reply:
The past dialogue was for simulating the example messages for the character the bot would be playing, which seemed to help a lot in enforcing a type of personality. This bot was originally created for gpt4xalpaca, but with newer models, maybe its not so necessary?